transaction/beef.go

```diff
@@ -803,115 +800,214 @@ func (b *Beef) Verify(ctx context.Context, chainTracker chaintracker.ChainTracke
+				} else {
+					needsValidation = append(needsValidation, beefTx)
+				}
 			}
 		}
 	}
 
-	// Try to validate any transactions whose inputs are now known
-	oldLen := -1
-	for oldLen != len(queue) {
-		oldLen = len(queue)
-		newQueue := make([]*BeefTx, 0, len(queue))
-		for _, beefTx := range queue {
+	// Iteratively validate transactions that depend on other transactions
+	for len(needsValidation) > 0 {
+		progress := false
+		var stillNeedsValidation []*BeefTx
+
+		for _, beefTx := range needsValidation {
+			// Check if all inputs are valid
+			allInputsValid := true
 			if beefTx.Transaction != nil {
-				allInputsValid := true
-				for _, in := range beefTx.Transaction.Inputs {
-					if !validTxids[in.SourceTXID.String()] {
+				for _, input := range beefTx.Transaction.Inputs {
+					if !validTxids[input.SourceTXID.String()] {
 						allInputsValid = false
 						break
 					}
 				}
-				if allInputsValid {
-					validTxids[beefTx.Transaction.TxID().String()] = true
-					res.Valid = append(res.Valid, beefTx.Transaction.TxID().String())
-				} else {
-					newQueue = append(newQueue, beefTx)
+			}
+
+			if allInputsValid {
+				txid := beefTx.Transaction.TxID().String()
+				validTxids[txid] = true
+				hasProof = append(hasProof, beefTx)
+				progress = true
+			} else {
+				stillNeedsValidation = append(stillNeedsValidation, beefTx)
+			}
+		}
+
+		needsValidation = stillNeedsValidation
+		if !progress {
+			// No progress made - remaining transactions are not valid
+			for _, beefTx := range needsValidation {
+				if beefTx.Transaction != nil {
+					result.NotValid = append(result.NotValid, beefTx.Transaction.TxID().String())
 				}
 			}
+			break
+		}
+	}
+
+	// Populate result lists
+	// Add transactions with missing inputs
+	for _, beefTx := range withMissingInputs {
+		if beefTx.Transaction != nil {
+			txid := beefTx.Transaction.TxID().String()
+			result.WithMissingInputs = append(result.WithMissingInputs, txid)
 		}
-		queue = newQueue
 	}
 
-	// Now, whatever is left in queue is not valid
-	for _, beefTx := range queue {
+	// Add txid-only transactions
+	for _, beefTx := range txidOnly {
+		var txid string
+		if beefTx.KnownTxID != nil {
+			txid = beefTx.KnownTxID.String()
+		} else if beefTx.Transaction != nil {
+			txid = beefTx.Transaction.TxID().String()
+		} else {
+			continue
+		}
+		result.TxidOnly = append(result.TxidOnly, txid)
+		if validTxids[txid] {
+			result.Valid = append(result.Valid, txid)
+		}
+	}
+
+	// Add valid transactions with proofs (in dependency order)
+	for _, beefTx := range hasProof {
 		if beefTx.Transaction != nil {
-			res.NotValid = append(res.NotValid, beefTx.Transaction.TxID().String())
+			txid := beefTx.Transaction.TxID().String()
+			result.Valid = append(result.Valid, txid)
 		}
 	}
 
-	for k := range missing {
-		res.MissingInputs = append(res.MissingInputs, k)
+	// Populate missing inputs list
+	for txid := range missingInputs {
```

Comment:
Similar comment ☝️ - preallocate
```suggestion
	result.MissingInputs = make([]string, 0, len(missingInputs))
	for txid := range missingInputs {
```